### ISSUE TITLE
Current Connection Max Out when using Distributed Schedule Service

### ISSUE THREAD

**Issue by yjain3326** [redisson/redisson#1087]
Hi,

We are using AWS Elastic Cache (3.2.4-cache.m3.xLarge). 
We are trying to solve for scheduling events using Redisson Executor Service (https://github.com/redisson/redisson/wiki/9.-distributed-services). We have a setup of 1master and 4 slave nodes.

When we are trying to send scheduling request at rate of around 1000 request per sec for more than 1hour duration, we are seeing that current connections shoots up to max and they are never reclaimed and the app starts throwing errors.

These are our Redisson Client setting: (Replicated servers configuration)
Elasticache node scan interval in milliseconds=5000
masterConnectionPoolSize=5000
slaveConnectionPoolSize=5000
masterConnectionMinimumIdleSize=50
slaveConnectionMinimumIdleSize=50
idleConnectionTimeout=30000
connectTimeout=30000
timeout=30000
Executor service workers=10

We are creating a executor service on startup and use it throughout the app. We are using the scheduleAsync function and Runnable implementation.

We have a 6node (EC2 instanceC3.xLarge) app Cluster taking to Redis. (Each node has the above configuration)

Can you please help provide guidance as to what might be the issue here and how we can resolve it.

Thanks.

**Comment by mrniko** [comment]
Which Redisson version do you use?

**Comment by yjain3326** [comment]
Hi,

We are using 3.5.3 redisson version.

**Comment by mrniko** [comment]
> the app starts throwing errors

Could you share the trace?

**Comment by yjain3326** [comment]
Basically we are seeing RedisTimeoutException.

` ERROR QueueTransferTask [] - Unable to send command: (EVAL) with params: [Ljava.lang.Object;@58cc9cb after 3 retry attempts
org.redisson.client.RedisTimeoutException: Unable to send command: (EVAL) with params: [Ljava.lang.Object;@58cc9cb after 3 retry attempts
	at org.redisson.command.CommandAsyncService$8.run(CommandAsyncService.java:570) [redisson-3.5.3.jar:?]
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663) [netty-common-4.1.15.Final.jar:?]
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738) [netty-common-4.1.15.Final.jar:?]
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466) [netty-common-4.1.15.Final.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]`

`ERROR RedissonRemoteService [] - Can't process the remote service request.
org.redisson.client.RedisTimeoutException: Unable to send command: (BLPOP) with params: [Ljava.lang.Object;@68a3182 after 3 retry attempts
	at org.redisson.command.CommandAsyncService$8.run(CommandAsyncService.java:570) [redisson-3.5.3.jar:?]
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663) [netty-common-4.1.15.Final.jar:?]
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738) [netty-common-4.1.15.Final.jar:?]
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466) [netty-common-4.1.15.Final.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]`

Also When I do an "INFO" command on the Redis master Node, I see all the connections under "blocked_clients"

**Comment by yjain3326** [comment]
I looked at logs and found these trace too:
` MasterConnectionPool [] - host *********:6379 disconnected due to failedAttempts=3 limit reached
io.netty.channel.AbstractChannel$AnnotatedSocketException: Cannot assign requested address: ********:6379
	at sun.nio.ch.Net.connect0(Native Method) ~[?:1.8.0_144]
	at sun.nio.ch.Net.connect(Net.java:454) ~[?:1.8.0_144]
	at sun.nio.ch.Net.connect(Net.java:446) ~[?:1.8.0_144]
	at sun.nio.ch.SocketChannelImpl.connect(SocketChannelImpl.java:648) ~[?:1.8.0_144]
	at io.netty.util.internal.SocketUtils$3.run(SocketUtils.java:83) ~[netty-common-4.1.15.Final.jar:4.1.15.Final]
	at io.netty.util.internal.SocketUtils$3.run(SocketUtils.java:80) ~[netty-common-4.1.15.Final.jar:4.1.15.Final]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_144]
	at io.netty.util.internal.SocketUtils.connect(SocketUtils.java:80) ~[netty-common-4.1.15.Final.jar:4.1.15.Final]
	at io.netty.channel.socket.nio.NioSocketChannel.doConnect(NioSocketChannel.java:338) ~[netty-transport-4.1.15.Final.jar:4.1.15.Final]
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.connect(AbstractNioChannel.java:254) ~[netty-transport-4.1.15.Final.jar:4.1.15.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.connect(DefaultChannelPipeline.java:1291) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:545) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:530) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:47) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:545) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:530) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:47) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:545) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:530) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:47) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:545) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:530) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:512) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.DefaultChannelPipeline.connect(DefaultChannelPipeline.java:994) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.channel.AbstractChannel.connect(AbstractChannel.java:259) ~[netty-transport-4.1.15.Final.jar:?]
	at io.netty.bootstrap.Bootstrap$3.run(Bootstrap.java:252) ~[netty-transport-4.1.15.Final.jar:4.1.15.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) [netty-common-4.1.15.Final.jar:?]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403) [netty-common-4.1.15.Final.jar:?]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463) [netty-transport-4.1.15.Final.jar:4.1.15.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) [netty-common-4.1.15.Final.jar:?]
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138) [netty-common-4.1.15.Final.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
Caused by: java.net.BindException: Cannot assign requested address
	... 32 more`

` QueueTransferTask [] - MasterConnectionPool no available Redis entries.  Disconnected hosts: [********:6379]
org.redisson.client.RedisConnectionException: MasterConnectionPool no available Redis entries.  Disconnected hosts: [*******:6379]
 at org.redisson.connection.pool.ConnectionPool.get(ConnectionPool.java:196) ~[redisson-3.5.3.jar:?]
 at org.redisson.connection.pool.MasterConnectionPool.get(MasterConnectionPool.java:31) ~[redisson-3.5.3.jar:?]
 at org.redisson.connection.MasterSlaveEntry.connectionWriteOp(MasterSlaveEntry.java:417) ~[redisson-3.5.3.jar:?]
 at org.redisson.connection.MasterSlaveConnectionManager.connectionWriteOp(MasterSlaveConnectionManager.java:706) ~[redisson-3.5.3.jar:?]
 at org.redisson.command.CommandAsyncService.async(CommandAsyncService.java:506) ~[redisson-3.5.3.jar:?]
 at org.redisson.command.CommandAsyncService$8.run(CommandAsyncService.java:585) [redisson-3.5.3.jar:?]
 at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663) [netty-common-4.1.15.Final.jar:?]
 at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738) [netty-common-4.1.15.Final.jar:?]
 at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466) [netty-common-4.1.15.Final.jar:?]
 at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]`

**Comment by mrniko** [comment]
> host *********:6379 disconnected due to failedAttempts=3 limit reached io.netty.channel.AbstractChannel$AnnotatedSocketException: Cannot assign requested address: ********:6379 

The cause of `Cannot assign requested address` error is that netty tries to open port for connection on client side which already bounded with another connection. Have you tried to reduce connections amount? I see it's 5000. That's to much for 10 workers

**Comment by yjain3326** [comment]
Yes I have tried with 1000 and 500 connections as well and have seen the same error.

**Comment by mrniko** [comment]
try set `masterConnectionMinimumIdleSize = masterConnectionPoolSize = 500` and `slaveConnectionMinimumIdleSize = slaveConnectionPoolSize = 500`

This initialize all connections during Redisson startup process

**Comment by yjain3326** [comment]
I tried changing this but with that I see a lot of these errors now:

`org.redisson.client.RedisTimeoutException: Unable to send command: (EVAL) with params: [Ljava.lang.Object;@292fdab2 after 3 retry attempts
	at org.redisson.command.CommandAsyncService$8.run(CommandAsyncService.java:570) [redisson-3.5.3.jar:?]
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663) [netty-common-4.1.15.Final.jar:?]
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738) [netty-common-4.1.15.Final.jar:?]
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466) [netty-common-4.1.15.Final.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]`

**Comment by mrniko** [comment]
Do you have any telemetry related to instance load? Can it be overloaded or something?

**Comment by jackygurui** [comment]
@yjain3326 What is netstat telling you about the connection states?

**Comment by yjain3326** [comment]
@mrniko CPU is under 25% and Freeable Memory is over 13.5GB for the Redis Instance. Any other metric you would like me to check?

**Comment by yjain3326** [comment]
@jackygurui netstat shows all connections to port 6379 as ESTABLISHED on the EC2 instances.
On the Redis instance I see all the connections under blocked_client. 
`# Clients
connected_clients:3010
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:3000`

**Comment by mrniko** [comment]
@yjain3326 

Could you check EC2 instances too?

**Comment by yjain3326** [comment]
@mrniko The avg CPU for EC2 instance is under 35% and Heap usage is under 25%

**Comment by yjain3326** [comment]
@mrniko For all the clients connected to my redis instance I see something like this:
` id=46237701 addr=***** fd=61 name= age=823 idle=823 flags=b db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=blpop`

Although I have set an idleTimeout for my redissonClient config, the connections are not getting Unblock.

**Comment by mrniko** [comment]
@yjain3326 

Connections used by BLPOP command are not released until blpop command execution is finished

**Comment by yjain3326** [comment]
@mrniko Could you provide any info on why that command is in stuck state and is not getting finished? Also, I think currently the timeout for this command is set to 24hours, is there a way to set it to a lower number?

This is my sample implementation:

` public boolean scheduleJob(Message message){
 try{
 RScheduledFuture<?> future = executor.schedule(new ScheduleExecutor(message.toJsonString()), delayInMillis, TimeUnit.MILLISECONDS);
 save(future.getTaskId());//saving in DB
 return true;
 }catch(Exception e){
 e.printStackTrace();
 }
 return false;
 }`

`public class ScheduleExecutor implements Runnable{

    @RInject
    private RedissonClient redissonClient;

    //Json Format Request to be sent to SQS
    private String request;

    public ScheduleExecutor(){
    }

    public ScheduleExecutor(String request){
        this.request = request;
    }

    public String getRequest() {
        return request;
    }

    public void setRequest(String request) {
        this.request = request;
    }

    @Override
    public void run(){
        try{
                    send(queueName, request); //sending to SQS Queue
                } catch (Exception e){
            e.printStackTrace();
        }
    }
}`

**Comment by mrniko** [comment]
@yjain3326 

> Could you provide any info on why that command is in stuck state and is not getting finished? 

BLPOP used in blocking queues. In context of ExecutorService it used to wait for a new task.

**Comment by yjain3326** [comment]
@mrniko I changed the redisson library version to 3.0.0 and the issue seems to have gone now.

**Comment by mrniko** [comment]
Can you write some test case to reproduce this?

**Comment by yjain3326** [comment]
@mrniko Executing this for about 60min should reproduce the issue:

`public class TestCase  implements Callable<String>{

    private int count;
    private RScheduledExecutorService redissonExecutorService;


    public TestCase(){
    }

    public TestCase(int count, RScheduledExecutorService redissonExecutorService ){
        this.count = count;
        this.redissonExecutorService = redissonExecutorService;
    }

    public String call(){
        for(int i = 0; i<count; i++){
            //3min delay
            long delayInMillis = System.currentTimeMillis() + (3*60*1000);
            RScheduledFuture<?> future = redissonExecutorService.scheduleAsync(new Executor("Payload"), delayInMillis, TimeUnit.MILLISECONDS);
            if (future != null && !StringUtils.isEmpty(future.getTaskId())) {
                System.out.print("Redis Task Id is: " + future.getTaskId());
            }
        }

        return null;
    }

    public static void main(String[] args) {

        String nodeList = "redis://localhost:6379";
        Config config = new Config();
        RedissonClient redissonClient = null;
        RScheduledExecutorService redissonExecutorService = null;
        int durationInMinutes = 30;
        int totalThreads = 10;
        int count = 100;

        if (nodeList != null) {
            StringTokenizer nodes = new StringTokenizer(nodeList, ",");

            while (nodes.hasMoreTokens()) {
                config.useReplicatedServers()
                        .addNodeAddress(nodes.nextToken());
            }
            config.useReplicatedServers().setScanInterval(5000);
            config.useReplicatedServers().setIdleConnectionTimeout(30000);
            config.useReplicatedServers().setMasterConnectionMinimumIdleSize(50);
            config.useReplicatedServers().setMasterConnectionPoolSize(250);
            config.useReplicatedServers().setSlaveConnectionMinimumIdleSize(50);
            config.useReplicatedServers().setSlaveConnectionPoolSize(250);
            config.useReplicatedServers().setConnectTimeout(30000);
            config.useReplicatedServers().setTimeout(30000);

            RedissonNodeConfig nodeCnfig = new RedissonNodeConfig(config);
            nodeCnfig.setExecutorServiceWorkers(Collections.singletonMap("EXECUTORSERVICENAME", 10));

            redissonClient = Redisson.create(config);
            RedissonNode node = RedissonNode.create(nodeCnfig, (Redisson) redissonClient);
            node.start();

            redissonExecutorService = redissonClient.getExecutorService("EXECUTORSERVICENAME");

            long currentTime = System.currentTimeMillis();
            Calendar time = Calendar.getInstance();
            time.add(Calendar.MINUTE, durationInMinutes);
            try {
                while (currentTime < time.getTimeInMillis()) {
                    ExecutorService service = Executors.newFixedThreadPool(totalThreads);
                    List<Future<String>> futures = new ArrayList<>();

                    try {
                        for (int i = 0; i < totalThreads; i++) {
                            futures.add(service.submit(new TestCase(count,redissonExecutorService)));
                        }
                        currentTime = System.currentTimeMillis();
                        Thread.sleep(1000);
                    }catch(Exception ex){
                        ex.printStackTrace();
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }

        }


    }
}`


`public class Executor implements Runnable{

    private String request;

    public Executor(){
    }

    public Executor(String request ){
        this.request = request;
    }

    @Override
    public void run(){

        try{

            System.out.print("Request is: "+request);
        }catch (Exception e){
            e.printStackTrace();
        }
    }
}`

**Comment by mrniko** [comment]
>         //3min delay
>        long delayInMillis = System.currentTimeMillis() + (3*60*1000);

Trying to run it. Why do you sum `System.currentTimeMillis()`? delay will be infinity in this case

**Comment by yjain3326** [comment]
My Bad, you can just use 3min as delay and run it.

**Comment by mrniko** [comment]
Yes I finally have reproduced it. It's a bug. Thanks for report

**Comment by yjain3326** [comment]
@mrniko Happy to help!

**Comment by mrniko** [comment]
Fixed

**Comment by yjain3326** [comment]
@mrniko is this released?

**Comment by mrniko** [comment]
@yjain3326 not yet

**Comment by mrniko** [comment]
@yjain3326 2.10.5 and 3.3.5 versions were released
